### PR TITLE
abci: ADR-008 Merge `abci-rs` into `tendermint-rs`

### PR DESCRIPTION
**Pull request by devashishdxt**
[Rendered](https://github.com/devashishdxt/tendermint-rs/blob/abci-rs-adr/docs/architecture/adr-008-abci-rs-merge.md)

### CODE REVIEW

File: docs/architecture/adr-008-abci-rs-merge.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# ADR 008: Merge `abci-rs` into `tendermint-rs`
+
+## Changelog
+* 2020-08-11: First draft
+
+## Context
+
+As mentioned [here](https://github.com/informalsystems/tendermint-rs/issues/388#issue-646627068), one high level goal
+for this `tendermint-rs` is to merge an already existing crate for build ABCI servers. There are three main crates for
+this purpose:
```
Comment (xla) [inline docs/architecture/adr-008-abci-rs-merge.md:10]:
```suggestion
As mentioned #388 one of the high level goals for the `tendermint-rs` repo is to merge an already existing crate to build ABCI servers with. There are three existing crates for this purpose:
```

File: docs/architecture/adr-008-abci-rs-merge.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+1. [`rust-abci`](https://github.com/tendermint/rust-abci)
+1. [`abci-rs`](https://github.com/devashishdxt/abci-rs)
+1. [`abci2`](https://github.com/nomic-io/abci2)
+
+### History
+
+There was a [proposal](https://github.com/tendermint/rust-abci/issues/61) for rearchitecturing `rust-abci` where a lot
```
Comment (xla) [inline docs/architecture/adr-008-abci-rs-merge.md:18]:
```suggestion
There was a [proposal](https://github.com/tendermint/rust-abci/issues/61) for rearchitecting `rust-abci` where a lot
```

File: docs/architecture/adr-008-abci-rs-merge.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+Sync`, also `abci-rs` is not easy to work with when rest of the application is using `async` Rust etc.). A decision was
+made to explore other architectures for `rust-abci` which are also compatible with `async` Rust.
+
+### Current state
+
+Currently, there are two new crates which were created in order to explore other architectures for `rust-abci`,
+`abci-rs` and `abci2`.
+
+`abci2`'s architecture is very minimal and is akin to a wrapper around raw `TcpStream` with `protobuf` encoding/decoding
+of ABCI types. All the burden of handing requests of different types in different ways is left for the developer of
```
Comment (xla) [inline docs/architecture/adr-008-abci-rs-merge.md:29]:
```suggestion
of ABCI types. All the burden of handing requests of different types in different ways is left to the developer.
```

File: docs/architecture/adr-008-abci-rs-merge.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ABCI server.
```
Comment (xla) [inline docs/architecture/adr-008-abci-rs-merge.md:30]:
```suggestion
```

File: docs/architecture/adr-008-abci-rs-merge.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+### Positive
+
+- `abci-rs` will be kept up to date with the `tendermint-rs` development and remain compatible with other related
+  crates.
+- `abci-rs` can share dependencies with other related crates (for example, simpler types from `proto`, etc.).
+- Less burden on developer/crate user.
+
+### Negative/Neutral
+
+- Developer is forced to use `async` Rust.
```
Comment (ebuchman) [inline docs/architecture/adr-008-abci-rs-merge.md:52]: I really like what is described in (1) here and seems to be a possible convergence point amidst all the discussion to work towards, and something that is done in other rust crates, where the core is implemented synchronously and an async wrapper is provided. The extra maintenance cost for the wrapper seems minimal compared to the user/flexibility benefit, especially if the synchronous core is simple enough. 

Would it make sense for abci2 to be used as such a core and adapt abci-rs to be an async wrapper around it, and expose both ?

**Pull request closed without merging**